### PR TITLE
Search tags in waterfall

### PR DESCRIPTION
**Pull request by rajdeepbharati**
## Contributor Checklist:

* [ ] I have updated the unit tests
* [ ] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by tardyp** [comment]
I would limit the number of displayed tags to 20. and add suspension points to indicate there is more, and one should use search

File: www/waterfall_view/src/module/waterfall.tpl.jade [hunk 1]
```diff
@@ -3,12 +3,16 @@
...
             p loading
-    p Tags: 
-        span(ng-show="w.all_tags.length > 0" ng-repeat="tag in w.all_tags")
+    p
+        form(role="search" style="width:150px")
+            input(type="text" ng-model="searchedTag" class="form-control" placeholder="Search for tags")
+    p
+        span.label.clickable.label-danger(ng-click="w.clearTags()") x
+        | &nbsp;
+        span(ng-show="w.all_tags.length > 0" ng-repeat="tag in w.all_tags | filter: searchedTag")
```
Comment (tardyp) [inline www/waterfall_view/src/module/waterfall.tpl.jade]: ng-show is redundant with ng-repeat.

**Comment by tardyp** [comment]
I think we should rather use a ready to use tag selector like this one:
http://mbenford.github.io/ngTagsInput/demos

I use it in buildbot travis already

**Comment by rajdeepbharati** [comment]
@tardyp Implementing the + - (and/or/none) filtering might be difficult if we use an external library. I'll try to do it and let you know

**Comment by tardyp** [comment]
I think we can have several tag inputs that you can create using a + button.
each tag within a tag input is "and" and each tag input groups are "or"

image

The tag-input is from the lib, and then you have a small directive to manage tag groups

**Comment by rajdeepbharati** [comment]
Okat, that sounds good (having multiple input fields).
> The tag-input is from the lib, and then you have a small directive to manage tag groups

Does directive mean custom directive?

**Comment by tardyp** [comment]
> Does directive mean custom directive?

Yes. those are quite simple. there are example of such object list editor in buildbot travis

https://github.com/buildbot/buildbot_travis/blob/master/buildbot_travis/ui/src/app/configEditors/projects.tpl.jade#L2
https://github.com/buildbot/buildbot_travis/blob/master/buildbot_travis/ui/src/app/configEditors/projects.tpl.jade#L78

**Comment by mojca** [comment]
Unrelated to the search itself, but I believe that when a tag selection is changed, waterfall should be redrawn from scratch. After applying a tag and while still having two builds available it seems like detective work to dig them out. (I can open a separate issue for that.)

Regarding tags themselves, even with this extra layer I'm still not sure if we could express the full boolean algebra, and some other expressions might still be relatively complicated to write (say: show any of those 10 ports built on any of three platforms; probably 30 queries).

**Comment by rajdeepbharati** [comment]
@mojca redrawing would probably be a bad thing to do.
> After applying a tag and while still having two builds available

Could you show an example of this?

**Comment by tardyp** [comment]
> Regarding tags themselves, even with this extra layer I'm still not sure if we could express the full boolean algebra, and some other expressions might still be relatively complicated to write (say: show any of those 10 ports built on any of three platforms; probably 30 queries).


Sounds like a relevent use case. 2 options:

- if we click on the "and" button, then the logic is inverted, and it becomes an "or"
- We implementing arbitrary logic with regex (i.e tag you specify can be regex)

**Comment by mojca** [comment]
> > After applying a tag and while still having two builds available
> 
> Could you show an example of this?

Here's an example of the lack of redraw:
* https://build.contextgarden.net/#/builders?tags=%2Bc%2Fluatex (works fine)
image
* https://build.contextgarden.net/#/waterfall?tags=c%2Fluatex (empty, some times on the axis are offset by one pixel)
image

But please note that this is something that probably deserves a separate issue to be opened.

**Pull request closed without merging**